### ISSUE TITLE
Computing perplexity as a metric: K.pow() doesn't work?!

### ISSUE THREAD

**Issue by DavidNemeskey** [keras-team/keras#8267]
I started reimplementing the PTB language model in [Zaremba (2014)](https://arxiv.org/abs/1409.2329) as an exercise, and implemented a perplexity metric following [icoxfog417's comment](https://github.com/fchollet/keras/issues/2317#issuecomment-307605458). However, the perplexity computed by Keras is incorrect. When running the code, I get the following output:

```
Train on 46460 samples, validate on 3680 samples
Epoch 1/13
0 {'size': 20, 'batch': 0}
0 {'perplexity': 592.40344, 'size': 20, 'loss': 9.2104359, 'crossentropy': 9.2104368, 'batch': 0}
   20/46460 [..............................] - ETA: 3:16:35 - loss: 9.2104 - crossentropy: 9.2104 - perplexity: 592.40341 {'size': 20, 'batch': 1}
1 {'perplexity': 590.44135, 'size': 20, 'loss': 9.2056398, 'crossentropy': 9.2056398, 'batch': 1}
   40/46460 [..............................] - ETA: 1:48:36 - loss: 9.2080 - crossentropy: 9.2080 - perplexity: 591.42242 {'size': 20, 'batch': 2}
2 {'perplexity': 579.80609, 'size': 20, 'loss': 9.1790438, 'crossentropy': 9.1790428, 'batch': 2}
   60/46460 [..............................] - ETA: 1:19:10 - loss: 9.1984 - crossentropy: 9.1984 - perplexity: 587.55033 {'size': 20, 'batch': 3}
3 {'perplexity': 474.20993, 'size': 20, 'loss': 8.8361568, 'crossentropy': 8.8361578, 'batch': 3}
   80/46460 [..............................] - ETA: 1:04:20 - loss: 9.1078 - crossentropy: 9.1078 - perplexity: 559.21524 {'size': 20, 'batch': 4}
4 {'perplexity': 340.86914, 'size': 20, 'loss': 7.9213591, 'crossentropy': 7.9213591, 'batch': 4}
  100/46460 [..............................] - ETA: 55:37 - loss: 8.8705 - crossentropy: 8.8705 - perplexity: 515.5460  5 {'size': 20, 'batch': 5}
5 {'perplexity': 349.80087, 'size': 20, 'loss': 7.2714424, 'crossentropy': 7.2714429, 'batch': 5}
  120/46460 [..............................] - ETA: 49:34 - loss: 8.6040 - crossentropy: 8.6040 - perplexity: 487.92186 {'size': 20, 'batch': 6}
6 {'perplexity': 359.06558, 'size': 20, 'loss': 6.9727578, 'crossentropy': 6.9727588, 'batch': 6}
  140/46460 [..............................] - ETA: 45:34 - loss: 8.3710 - crossentropy: 8.3710 - perplexity: 469.51387 {'size': 20, 'batch': 7}
7 {'perplexity': 421.70245, 'size': 20, 'loss': 6.9462433, 'crossentropy': 6.9462428, 'batch': 7}
Logs: {'perplexity': 421.70245, 'size': 20, 'loss': 6.9462433, 'crossentropy': 6.9462428, 'batch': 7}
  160/46460 [..............................] - ETA: 42:24 - loss: 8.1929 - crossentropy: 8.1929 - perplexity: 463.53748 {'size': 20, 'batch': 8}
8 {'perplexity': 385.3887, 'size': 20, 'loss': 6.6696272, 'crossentropy': 6.6696272, 'batch': 8}
Logs: {'perplexity': 385.3887, 'size': 20, 'loss': 6.6696272, 'crossentropy': 6.6696272, 'batch': 8}
  180/46460 [..............................] - ETA: 39:56 - loss: 8.0236 - crossentropy: 8.0236 - perplexity: 454.85429 {'size': 20, 'batch': 9}
9 {'perplexity': 445.72867, 'size': 20, 'loss': 6.8643327, 'crossentropy': 6.8643322, 'batch': 9}
Logs: {'perplexity': 445.72867, 'size': 20, 'loss': 6.8643327, 'crossentropy': 6.8643322, 'batch': 9}
  200/46460 [..............................] - ETA: 38:05 - loss: 7.9077 - crossentropy: 7.9077 - perplexity: 453.941610 {'size': 20, 'batch': 10}
```

I log the `logs` dictionary after each batch (the lines above the progbar output), because progbar averages the metrics. I also added another metric, _crossentropy_ to account for the first part of the computation. It is not always the same as the loss (which is interesting in itself), but close enough, so that part seems to be OK.

However, it is clear that the perplexity value is slipping:
- after the first batch, `2 ^ 9.2104359 == 592.403`
- after the last: `2 ^ 6.8643327 == 116.512 != 445.72867`

So the error might be in two places:
- `K.pow`: however, it is just a call `tf.pow`, and both seem to function fine when called in isolation
- maybe something affects the perplexity calculation (another form of averaging? But why doesn't it affect the cross entropy metric, then?

My code is:
```Python
#!/usr/bin/env python3
# vim: set fileencoding=utf-8 :

import numpy as np

import keras.backend as K
from keras.callbacks import LambdaCallback
from keras.initializers import RandomUniform
from keras.layers import Dense, Dropout
from keras.layers import Embedding
from keras.layers import LSTM, TimeDistributed
from keras.models import Model, Sequential


def perplexity(y_true, y_pred):
    """
    The perplexity metric. Why isn't this part of Keras yet?!

    BTW doesn't really work.
    """
    cross_entropy = K.sparse_categorical_crossentropy(y_true, y_pred)
    perplexity = K.pow(2.0, cross_entropy)
    return perplexity


def crossentropy(y_true, y_pred):
    return K.sparse_categorical_crossentropy(y_true, y_pred)


def build_model(config):
    vocab_len = 10000
    hidden_size = 200
    timesteps = 20
    batch_size = 20
    initializer = RandomUniform(-0.1, 0.1)
    model = Sequential()
    model.add(Embedding(vocab_len, hidden_size,
                        embeddings_initializer=initializer,
                        batch_input_shape=(batch_size, timesteps),
                        input_length=timesteps))
    model.add(LSTM(hidden_size, unit_forget_bias=True, stateful=True,
                   unroll=True, return_sequences=True))
    model.add(LSTM(hidden_size, unit_forget_bias=True, stateful=True,
                   unroll=True, return_sequences=True))
    model.add(TimeDistributed(Dense(vocab_len, activation='softmax')))

    model.compile(loss='sparse_categorical_crossentropy',
                  optimizer='rmsprop', metrics=[crossentropy, perplexity])
    return model


def read_data_file(file_name, batch_size, timesteps):
    data = np.load(file_name)['data']
    batch_len = data.size // batch_size
    batch_len = batch_len // timesteps * timesteps
    if batch_size * batch_len == data.size:
        batch_len -= timesteps
    data = np.resize(data, batch_size * (batch_len + 1)).reshape(batch_size, -1)
    return data[:, :-1].T, np.expand_dims(data[:, 1:].T, -1)


def read_data(train_file, valid_file, test_file, batch_size, timesteps):
    train = read_data_file(train_file, batch_size, timesteps)
    valid = read_data_file(valid_file, batch_size, timesteps)
    test = read_data_file(test_file, batch_size, timesteps)
    return train, valid, test


def main():
    model = build_model(None)
    (trainx, trainy), (validx, validy), (testx, testy) = read_data(
        'traini.npz', 'validi.npz', 'testi.npz', 20, 20)
    lc = LambdaCallback(on_batch_begin=lambda b, l: print(b, l),
                        on_batch_end=lambda b, l: print(b, l))
    model.fit(trainx, trainy, batch_size=20, epochs=13, shuffle=False,
              validation_data=(validx, validy), callbacks=[lc])


if __name__ == '__main__':
    main()
```

My environment:
- Python 3.5.2
- Tensorflow (CPU) 1.4rc1
- Keras latest master `3b15ee6954b688b9490fd885d628939fd35f76ca`

**Comment by DavidNemeskey** [comment]
I can provide the data files if needed.

**Comment by bolunwang** [comment]
Not an expert. Could it be because how metrics are averaged over samples?

avg(CE) != avg(2^CE)

**Comment by MohameddSheriff** [comment]
@DavidNemeskey have you found any solution?

**Comment by DavidNemeskey** [comment]
@MohameddSheriff Yes, and it's called Pytorch. :grinning: @bolunwang might have a point, but as I remember, I tried the formula outside the metric context and it worked. But don't quote me on that.

**Comment by ajayrfhp** [comment]
Could we just do `2**cross_entropy` ? The code seems to run.

**Comment by manharwalia11** [comment]
> @MohameddSheriff Yes, and it's called Pytorch. 😀 @bolunwang might have a point, but as I remember, I tried the formula outside the metric context and it worked. But don't quote me on that.

Can you please share the solution here.?

**Comment by Rohianon** [comment]
What was the final solution to this?
